rfcs/0049-flakes.md

```diff
@@ -0,0 +1,496 @@
+  set. The attributes can have arbitrary values; however, some tools
+  may require specific attributes to have a specific value (e.g. the
+  `nix` command may expect the value of `packages.x86_64-linux` to be
+  an attribute set of derivations built for the `x86_64-linux`
+  platform).
+
+## Flake inputs
+
+The attribute `inputs` specifies the dependencies of a flake. These
+specify the location of the dependency, or a symbolic flake identifier
+that is looked up in a registry or in a command-line flag. For
+example, the following specifies a dependency on the Nixpkgs and import-cargo
+repositories:
+
+    # A GitHub repository.
+    inputs.import-cargo = {
+      type = "github";
+      owner = "edolstra";
+      repo = "import-cargo";
+    };
+
+    # An indirection through the flake registry.
+    inputs.nixpkgs = {
+      type = "indirect";
+      id = "nixpkgs";
+    };
+
+Each input is fetched, evaluated and passed to the `outputs` function
+as a set of attributes with the same name as the corresponding
+input. The special input named `self` refers to the outputs and source
+tree of *this* flake. Thus, a typical `outputs` function looks like
+this:
+
+    outputs = { self, nixpkgs, import-cargo }: {
+      ... outputs ...
+    };
+
+It is also possible to omit inputs entirely and *only* list them as
+expected function arguments in `outputs`. Thus,
+
+    outputs = { self, nixpkgs }: ...;
+
+without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in
+the flake registry.
+
+Repositories that don't contain a `flake.nix` can also be used as
+inputs, by setting the input's `flake` attribute to `false`:
+
+    inputs.grcov = {
+      type = "github";
+      owner = "mozilla";
+      repo = "grcov";
+      flake = false;
+    };
+
+    outputs = { self, nixpkgs, grcov }: {
+      packages.x86_64-linux.grcov = stdenv.mkDerivation {
+        src = grcov;
+        ...
+      };
+    };
+
+The following input types are specified at present:
+
+* `git`: A Git repository or dirty local working tree.
+
+* `github`: A more efficient scheme to fetch repositories from GitHub
+  as tarballs. These have slightly different semantics from `git`
+  (in particular, the `revCount` attribute is not available).
+
+* `gitlab`: Like `github`, but for Git repositories hosted on GitLab.
+
+* `tarball`: A `.tar.{gz,xz,bz2}` file.
+
+* `path`: A directory in the file system. This generally should be
+  avoided in favor of `git` inputs, since `path` inputs have no
+  concept of revisions (only a content hash) or tracked files
+  (anything in the source directory is copied).
+
+* `hg`: A Mercurial repository.
+
+Transitive inputs can be overriden from a `flake.nix` file. For
+example, the following overrides the `nixpkgs` input of the `nixops`
+input:
+
+    inputs.nixops.inputs.nixpkgs = {
+      type = "github";
+      owner = "my-org";
+      repo = "nixpkgs";
+    };
+
+It is also possible to "inherit" an input from another input. This is
```

Comment:
I have a usability concern here:

If I understand it correctly, the default behaviour is to use the lock file of each flake to resolve its inputs. This would potentially lead to many different nixpkgs versions being used, creating long build times and large output closures with many duplicate files. It would also enable old versions of libraries to sneak in, creating a potential security concern for users who are not very familiar with Nix and use the default behaviour.
Being able to reproduce the same result is great, but if the same version is always reproduced by default, what is the cost for that?

Or is the lock file of the top level flake used to provide the inputs for all dependency flakes recursively?

Reply:
Yes, there can be many versions of nixpkgs in the flake input closure. However, if you're using Nixpkgs overlays or NixOS modules, only the top-level nixpkgs version matters, since overlays/modules get composed with the top-level nixpkgs. If needed you can override transitive inputs to point to the same version, e.g.
```
  inputs.nixpkgs.url = "nixpkgs/nixos-20.03";
  inputs.dwarffs.inputs.nixpkgs.follows = "nixpkgs";
  inputs.nix.inputs.nixpkgs.follows = "nixpkgs";
  inputs.blender-bin.inputs.nixpkgs.follows = "nixpkgs";
```

Reply:
This answers my question. If everyone uses overlays, this will not be an issue. It will be important to communicate this as the preferred way of providing package outputs in documentation. I will raise concerns again if I think this is not sufficiently communicated.

Reply:
I see a usability problem with composing independent overlays. This can easily lead to hard to debug recursion errors or conflicts if they override the same packages in different ways. For that reason I decided to not use overlays in NUR but having namespaces not conflicting with each other.

Reply:
> Yes, there can be many versions of nixpkgs in the flake input closure. However, if you're using Nixpkgs overlays or NixOS modules, only the top-level nixpkgs version matters, since overlays/modules get composed with the top-level nixpkgs. If needed you can override transitive inputs to point to the same version, e.g.
> 
> ```
>   inputs.nixpkgs.url = "nixpkgs/nixos-20.03";
>   inputs.dwarffs.inputs.nixpkgs.follows = "nixpkgs";
>   inputs.nix.inputs.nixpkgs.follows = "nixpkgs";
>   inputs.blender-bin.inputs.nixpkgs.follows = "nixpkgs";
> ```

How does this work if `dwarffs` has a dependency flake that depends on nixpkgs?
Like this?

```
   inputs.dwarffs.dependency.inputs.nixpkgs.follows = "nixpkgs";
```

Or is inputs flattened?

Reply:
> Note that even if only modules/overlays are used in the current implementation it downloads
nixpkgs for those flakes anyways also its are never used

No, flake inputs are fetched lazily, so if you don't use the `nixpkgs` input of the `dwarffs` input, it won't be fetched.